README
PyAuthenNTLM2 is a module that enables Apache to carry out authentication by
means of the NTLM protocol and a separate Domain Controller or Active
Directory server.

Introduction
============

The specific use case of PyAuthenNTLM2 is the following:

 * Users have valid accounts in an existing Windows domain, and you want such
   credentials to be used for HTTP authentication.
 * Apache runs on a server that is not part of the Windows domain.
 * Some of the users cannot use Kerberos. For instance, they connect via a proxy
   or they use machines that are not part of the domain. However, their client
   supports NTLM (all browsers do nowadays).
 * Some of the users can only use Basic, not even NTLM (possibly because they
   don't use a browser).
 * [EXPERIMENTAL] Only users that belong to certain groups are authorized.
   The groups are defined on the Active Directory server.

PyAuthenNTLM2 allows clients to authenticate via two schemes: NTLM or Basic.

As for several other authorization modules, PyAuthenNTLM will pass the username
to the underlying CGI application or webservice via the REMOTE_USER variable.
The user name is the Windows logon identifier, without the domain part
(that is, 'jdoe' and not 'DOMAIN\jdoe' or 'John Doe').

Security
--------

Although Kerberos (including its Microsoft variant "Integrated Windows
Authentication") is the best option in terms of security for HTTP
authentication, the venerable NTLM protocol is still a good compromise, much
better than the widespread Basic protocol and marginally better than Digest.
With Basic, anybody can pick up your password from the messages sent to the
server (unless you use TLS), whereas with Digest the web server needs local
access to the actual password. With NTLM, the webserver can relay every
authorization requests to the (Domain Controller or Active Directory server):
it never has to neither see nor store the real passwords.

A typical message flow is the following. First, the server (Apache)
refuses any request on a new connection:

Client ----[GET]---> Apache
Client <---[401]---- Apache , WWW-Authenticate: NTLM

The client crafts a first NTLM message to negotiate the authentication
parameters. Such message (Type 1) is relayed straight to the Domain Controller:

Client ----[GET]---> Apache , Authorization: TlRMT... (Type 1 message)
                     Apache ----[Type 1]---> DC

The Domain Controller generates a random challenge, completes the negotiation,
and sends the message. The message (Type 2) is relayed to the client:

                     Apache <---[Type 2]---- DC
Client <---[401]---- Server , WWW-Authenticate: NTLM TlRMT... (Type 2 message)

The client presents the popup dialog to the user to acquire the credentials.
The password is only used to encrypt the challenge. The resulting message
(Type 3) is relayed to the Domain Controller, which will communicate to the
server if the user is autentic or not:

Client ----[GET]---> Apache , Authorization: TlRMT... (Type 3 message)
                     Apache ----[Type 3]---> DC
                     Apache <---[Result]---- DC
Client <---[200]---- Apache

The password is never disclosed by the client. However, several versions of
NTLM exist. NTLMv1 is based on cryptography that by today standards is very
weak. That means that the password can be still brute forced from the
encrypted challenge. In order to avoid that, ensure that your clients
always use NTLMv2 [6].

PyAuthenNTLM2 handles equally well all the various NTLM variants (such as
NTLMv2 Session, and NTLMv2).

Requirements
============

* Mod-python [4]
* Python 2.x
* Apache 2.x
* PyCrypto 2.x (not required, but Basic auth will not be supported without it)

Test was carried out with:
 - a server with mod-python 3.3.1, Apache 2.2, and Python 2.6. 
 - clients with Internet Explorer 7/8, Firefox 7.0, and Chrome.
 - Windows 2003 Domain Controller and Active Directory server

Installation
============

python setup.py install

Usage
=====
Usage is best shown by an example of Apache configuration:

<Directory /var/lib/some_directory>

    AuthType NTLM
    AuthName WDOMAIN
    require valid-user

    PythonAuthenHandler pyntlm
    PythonOption Domain WDOMAIN
    PythonOption PDC 192.1.2.45
    PythonOption BDC 192.1.2.46

    # Bypass authentication for local clients.
    # Comment these lines if they should authenticate too.
    Order deny,allow
    Deny  from all
    Allow from 127.0.0.1
    Satify any

</Directory>

All non-local clients trying to access a URI mapped under the directory
/var/lib/some_directory will be asked for credentials valid in the Windows
Domain WDOMAIN. The user name to enter need to be in the format:
wdomain\useridentifier (for instance windom\jbrown).

Local clients (that is, those connecting from 127.0.0.1) will not be presented
with any request for authentication.

The following options exist:

* AuthType NTLM                     Always specify it like this.

* Require valid-user                Always specify it like this.

* Require user XYZ,WTY              Grants access only to users named XYZ or WTY.
                                    Multiple "Require user" option lines can be
                                    specified.

* AuthName [domain]                 Replace [domain] with the domain name to
                                    present to users in the pop-up dialog.

* PythonAuthenHandler [pyntlm.py]   Replace [pyntlm.py] with the complete,
                                    absolute path to the pyntlm.py script.

* PythonOption Domain [domain]      Replace [domain] with the Windows domain
                                    name (uppercase).

* PythonOption PDC [pdc]            Replace [pdc] with the address of the
                                    Primary Domain Controller (either IP or DNS
                                    name).

* PythonOption BDC [bdc]            Replace [bdc] with the address of the Backup
                                    Domain Controller (either IP or DNS name).
                                    This entry is optional.

Apache needs to be configured to send keep alives (directive 'KeepAlive On').

For SSL-protected sites, comment out the statement:

    SetEnvIf User-Agent ".*MSIE.*" \
             nokeepalive ssl-unclean-shutdown \
             downgrade-1.0 force-response-1.0

from the default Apache virtual host, or Internet Explorer will not manage to
make use of NTLM.

Experimental
------------
The [pdc]/[bdc] settings may also refer to an Active Directory server.
The syntax becomes slightly more complex:

    ldap://server[/baseDN]

where 'server' is the IP or DNS name of the Active Directory server, and the
optional 'baseDN' is the base Distinguished Name for the queries (only needed
for authorization, see below). For instance:

    ldap://10.12.13.1/DC=nasa,DC=gov

or equivalently (but more in compliance to RFC4516):

    ldap://10.12.13.1/DC%3Dnasa%2CDC%3Dgov

When using an Active Directory server, it is also possible to check if the
user is authorized to access page. More precisely, it is possible to grant
access only if the user is member of one group.

The option "Require group" can be used to pass the comma-separated list of groups
the user must belong to. The group identifier is the logon name, that is,
the 'sAMAccountName' attribute in Active Directory.

For instance, if you specified: 

<Directory /var/lib/some_directory>
    [ ... ]
    Require group Administrators,Power Users
    [ ... ]
</Directory>

Only users that belong to Administrators or Power Users will be granted access.
A user that does not belong to any of such groups is denied access, even if
its credentials were correct, unless its name is included in a "Require user"
option.

Caching
-------

NTLM is a protocol that authenticates **TCP connections**, not the individual
request like Basic or Digest. PyAuthenNTLM does not cache successful
autentications. Every time a new connection is established, you will see a query
to the Domain Controller or Active Directory from Apache. Browsers will
typically open several connections in parallel. Additionally, some browsers
may also trigger re-autentication within an established connection (e.g. Internet
Explorer for POST requests).

However, at the client side, browsers will do cache NTLM credentials, which
won't be asked again to ther user after the first time (which is the whole
point of Single Sign On!).

PyAuthenNTLM2 will cache positive group membership for 3 hours. That means that
if you remove a user from a group, such user will still be able to access
for up to 3 hours, unless you restart Apache. On the other hand, if you **add**
the user to a group, access should be granted immediately.

Logging
=======

All logs will show up in the Apache log file.

Troubleshooting
===============

Check list:

* Restart Apache each time you modify its configuration.
* Ensure that KeepAlive is On.
* Verify with various browser brands and versions, not just with one.
* If use SSL and cannot access using Internet Explorer but other browsers work,
  ensure that the "User-Agent MSIE" setting (see Usage above) is commented out
  in your site configuration.
* Increase the level of verbosity for the Apache log up to "Info". Note that 
  the LogLevel may be specified in multiple places in the Apache configuration.
  Ensure you are not setting the log level too high in the directory hierarchy.
* Ensure that mod-python is installed and activated. In the log file you should
  see messages like this:

[notice] mod_python: Creating 8 session mutexes based on 150 max processes and 0 max threads.

* Ensure that mod-python can find pyntlm.py. You should be able to see the
  following line for each a request, if the log level is set to "Info".

[info] [client 127.0.0.1] PYNTLM: Handling connection 0x0 from address 127.0.0.1 for GET URI /mysite/request

* If you are using group authorization, try without it, so that all users with
  a valid account can access the pages.
* Ensure that the authentication code is compatible with your Domain Controller,
  at least if you use a DC. Use the ntlm_client.py utility with the same
  settings from the Apache configuration:

python ntlm_client.py -u johndoe -p xxxxx -d DOMAINX -a 10.11.12.13 

  If you use an Active Directory server:

python ntlm_client.py -u johndoe -p xxxxx -d DOMAINX -a ldap://10.11.12.13 

  You should see the message:

User DOMAINX\johndoe was authenticated.

* Ensure that the browser is not using a configuration incompatible with
  the module. Use the http_client.py utility with the same settings
  from the Apache configuration. In case of NTLM autentication:

python http_client.py -s johndoe -p xxxxx -d DOMAINX http://apachesite/path/to/page

  And in case of Basic authentication:

python http_client.py -s johndoe -p xxxxx http://apachesite/path/to/page

  In either case you should see the message:

OK.

* Ensure you are not using group authorization check with a Domain Controller;
  that is not currently supported. It only works with an Active Directory server.
* If you are using group authorization check with an ActiveDirectory server,
  ensure that the base DN is correct in the Apache configuration file. See [7].
* If you are using group authorization check with an ActiveDirectory server,
  check if LDAP signing is required in the server. That is not currently
  supported by this module.
* Ensure that the group authorization code is compatible with your Active
  Directory server. Use the ntlm_client.py utility with the same settings
  from the Apache configuration:

python ntlm_client.py -u johndoe -p xxxxx -d DOMAINX -g Administrators -a ldap://10.11.12.13/DC=nasa,DC=gov

  You should see both 2 messages:

User DOMAINX\johndoe was authenticated.
User belongs to at least one group.

* If you have a problem of membership with a specific user you don't have the password
  for, you can explicit check what ntlm_client finds for that user, while authenticating
  with another accunt:

python ntlm_client.py -u otheraccount -p xxxxx -d DOMAINX -g Administrators -a ldap://10.11.12.13/DC=nasa,DC=gov -m johndoe -v

  You should see plenty of messages with the various DNs of the groups the user johndoe
  is found to be member of.

Thanks
======

Microsoft for the large amount of technical specifications about NTLM and SMB
it disclosed [1].
Eric Glass for his long article about NTLM [2]. In several ways, it is more
complete and precise than [1].
Gerald Ritcher and Shannon Eric Peevey for AuthenNTLM [3], which inspired this
module.

Contacts
========

Send an email to Legrandin <gooksankoo@hoiptorrow.mailexpire.com> or drop a
message at https://github.com/Legrandin/PyAuthenNTLM2.

References
==========

[1] http://msdn.microsoft.com/en-us/library/gg258393%28v=PROT.13%29.aspx
[2] http://davenport.sourceforge.net/ntlm.html
[3] http://search.cpan.org/~speeves/Apache2-AuthenNTLM-0.02/AuthenNTLM.pm
[4] http://www.modpython.org
[5] http://technet.microsoft.com/en-us/magazine/2006.08.securitywatch.aspx
[6] http://support.microsoft.com/kb/239869
[7] http://roadzy.blogspot.com/2011/02/finding-your-base-dn-in-active.html


Changelog
== Changelog

-- Version 2.2, 25 March 2012
* Added support for Active Directory.
* Added support for user-based authorization.
* Added (experimental) support for group-based authorization (AD only).
* Rearranged all code in a package (PyAuthNTLM2).
* Rearranged README in reStructuredText format (README.rst).
* Added Troubleshooting section to README.rst .

-- Version 2.1, 6 March 2012
* Added support for Basic authentication (thanks gmazzo!).
* Added two scripts for testing without a browser ({http|ntlm}_client.py).
* More details in the documentation.
* General clean up and better error handling.

-- Version 2.0, 20 December 2011
* First release



pyntlm.py
#!/usr/bin/env python
#
# PyAuthenNTLM2: A mod-python module for Apache that carries out NTLM authentication
#
# pyntlm.py
#
# Copyright 2011 Legrandin <gooksankoo@hoiptorrow.mailexpire.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys
import base64
import time
import urllib
from struct import unpack
from threading import Lock
from binascii import hexlify
from urlparse import urlparse

from mod_python import apache
from PyAuthenNTLM2.ntlm_dc_proxy import NTLM_DC_Proxy
from PyAuthenNTLM2.ntlm_ad_proxy import NTLM_AD_Proxy

use_basic_auth = True
try:
    from ntlm_client import NTLM_Client
except ImportError:
    use_basic_auth = False

#
# A connection can be in one of the following states when a request arrives:
#
# 1 Freshly opened: no authentication step has taken place yet.
#   req.connection.notes does not contain the key 'NTLM_AUTHORIZED' and
#   the cache does not contain any tuple under the connection's id.
#
# 2 Pending authentication: we sent the NTLM challenge to the client, and we 
#   are waiting for the response. req.connection.notes does not contain the
#   the key 'NTLM_AUTHORIZED' but the cache contains one tuple (NTLM_Proxy, timestamp)
#   under the connection's id.
#
# 3 Authenticated: all steps completed successfully. 
#   req.connection.notes contains the key 'NTLM_AUTHORIZED' (username) or
#   'BASIC_AUTHORIZED' (username and a password).
#   The cache should not contain any tuple under the connection's id.
#
# Since connections may be interrupted before receiving the challenge, objects older
# than 60 seconds are removed from cache when we have the chance.

class CacheConnections:

    def __init__(self):
        self._mutex = Lock()
        self._cache = {}

    def __len__(self):
        return len(self._cache)

    def remove(self, id):
        self._mutex.acquire()
        (proxy, ts) = self._cache.get(id, (None,None))
        if proxy:
            proxy.close()
            del self._cache[id]
        self._mutex.release()

    def add(self, id, proxy):
        self._mutex.acquire()
        self._cache[id] = ( proxy, int(time.time()) )
        self._mutex.release()
    
    def clean(self):
        now = int(time.time())
        self._mutex.acquire()
        for id, conn in self._cache.items():
            if conn[1]+60<now:
                conn[0].close()
                del self._cache[id]
        self._mutex.release()

    def has_key(self,id):
        return self._cache.has_key(id)

    def get_proxy(self, id):
        self._mutex.acquire()
        proxy = self._cache[id][0]
        self._mutex.release()
        return proxy

class CacheGroups:

    def __init__(self):
        self._mutex = Lock()
        self._cache = {}

    def __len__(self):
        return len(self._cache)

    def add(self, group, user):
        self._mutex.acquire()
        if not self._cache.has_key(group):
            self._cache[group]={}
        self._cache[group][user]=int(time.time())
        self._mutex.release()
    
    def clean(self):
        now = int(time.time())
        self._mutex.acquire()
        old = []
        for group, members in self._cache.items():
            for user in members:
                if members[user]+3*60*60<now:
                    old.append((group,user))
        for group, user in old:
            del self._cache[group][user]
        self._mutex.release()

    def has(self, group, user):
        if not self._cache.has_key(group):
            return False
        return self._cache[group].has_key(user)

cache = CacheConnections()
cacheGroups = CacheGroups()

def ntlm_message_type(msg):
    if not msg.startswith('NTLMSSP\x00') or len(msg)<12:
        raise RuntimeError("Not a valid NTLM message: '%s'" % hexlify(msg))
    msg_type = unpack('<I', msg[8:8+4])[0]
    if msg_type not in (1,2,3):
        raise RuntimeError("Incorrect NTLM message Type: %d" % msg_type)
    return msg_type

def parse_ntlm_authenticate(msg):
    '''Parse a Type3 NTLM message (binary form, not encoded in Base64).

    @return a tuple (username, domain)
    '''
    
    NTLMSSP_NEGOTIATE_UNICODE = 0x00000001
    idx = 28
    length, offset = unpack('<HxxI', msg[idx:idx+8])
    domain = msg[offset:offset+length]
    idx += 8
    length, offset = unpack('<HxxI', msg[idx:idx+8])
    username = msg[offset:offset+length]
    idx += 24
    flags = unpack('<I', msg[idx:idx+4])[0]
    if flags & NTLMSSP_NEGOTIATE_UNICODE:
        domain = str(domain.decode('utf-16-le'))
        username = str(username.decode('utf-16-le'))
    return username, domain

def decode_http_authorization_header(auth):
    '''Return a tuple with the parsed content of an HTTP Authorization header

    In case of NTLM, the first item is 'NTLM' and the second the Type 1 challenge.
    In case of Basic, the first item is 'Basic', the second the user name,
    and the third the password.

    In case of error, False is returned'''
    ah = auth.split(' ')
    if len(ah)==2:
        b64 = base64.b64decode(ah[1])
        if ah[0]=='NTLM':
            return ('NTLM', b64)
        elif ah[0]=='Basic' and use_basic_auth:
            (user, password) = b64.split(':')
            return ('Basic', user, password)
    return False

def handle_unauthorized(req):
    '''Prepare the correct HTTP headers for a 401 response.

    @return     The Apache return code for 401 response.
    '''

    req.err_headers_out.add('WWW-Authenticate', 'NTLM')
    if use_basic_auth:
        req.err_headers_out.add('WWW-Authenticate', 'Basic realm="%s"' % req.auth_name())
    req.err_headers_out.add('Connection', 'close')
    return apache.HTTP_UNAUTHORIZED

def connect_to_proxy(req, type1):
    '''Try to sequentially connect to all Domain Controllers in the list
    until one is available and can handle the NTLM transaction.

    @return A tuple with a NTLM_Proxy object and a NTLM challenge (Type 2).'''

    # Get configuration entries in Apache file
    try:
        domain = req.get_options()['Domain']
        pdc = req.get_options()['PDC']
        bdc = req.get_options().get('BDC', False)
    except KeyError, e:
        req.log_error('PYNTLM: Incorrect configuration for pyntlm = %s' % str(e), apache.APLOG_CRIT)
        raise
    ntlm_challenge = None
    for server in (pdc, bdc):
        if not server: continue
        try:
            if server.startswith('ldap:'):
                url = urlparse(server)
                decoded_path =urllib.unquote(url.path)[1:]
                req.log_error('PYTNLM: Initiating connection to Active Directory server %s (domain %s) using base DN "%s".' %
                    (url.netloc, domain, decoded_path), apache.APLOG_INFO)
                proxy = NTLM_AD_Proxy(url.netloc, domain, base=decoded_path)
            else:
                req.log_error('PYTNLM: Initiating connection to Domain Controller server %s (domain %s).' %
                    (server, domain), apache.APLOG_INFO)
                proxy = NTLM_DC_Proxy(server, domain)
            ntlm_challenge = proxy.negotiate(type1)
        except Exception, e:
            req.log_error('PYNTLM: Error when retrieving Type 2 message from server(%s) = %s' % (server,str(e)), apache.APLOG_CRIT)
        if ntlm_challenge: break
        proxy.close()
    else:
        raise RuntimeError("None of the Domain Controllers are available.")
    return (proxy, ntlm_challenge)

def handle_type1(req, ntlm_message):
    '''Handle a Type1 NTLM message. Send it to the Domain Controller
    and get back the challenge (the Type2 NTLM message that is).

    @req            The request that carried the message
    @ntlm_message   The actual Type1 message, in binary format
    '''
    cache.remove(req.connection.id)
    cache.clean()

    try:
        (proxy, ntlm_challenge) = connect_to_proxy(req, ntlm_message)
    except Exception, e:
        return apache.HTTP_INTERNAL_SERVER_ERROR

    cache.add(req.connection.id, proxy)
    req.err_headers_out.add('WWW-Authenticate', "NTLM " + base64.b64encode(ntlm_challenge))
    return apache.HTTP_UNAUTHORIZED

def check_authorization(req, proxy):
    '''Check if a user that was already authenticated by some previous steps
    is also authorized.

    Authorization is granted depending on the following Apache directives:

    Require valid-user must always be present
    Require user XYZ   authorizes any user named XYZ.
    Require group WER  authorizes any user which is member of the group WER.
                       Group membership is checked at the Active Directory
                       server.
    
    Multiple users and groups can be specified on the same Require line,
    provided they are separated by a comma.

    @req        The request for which authentication was successful (req.user exists).
    @proxy      The proxy that keeps membership data.
    @return     True if the user is authorized, False otherwise.
    '''
   
    rules = ''.join(req.requires()).strip()
    if rules=='valid-user' or cacheGroups.has(rules, req.user):
        return True
    groups = []
    for r in req.requires():
        if r.lower().startswith("user "):
            users = [ u.strip() for u in r[5:].split(",")]
            if req.user in users:
                req.log_error('PYNTLM: Authorization succeeded for user %s and URI %s.' %
                    (req.user,req.unparsed_uri), apache.APLOG_INFO)
                return True
        if r.lower().startswith("group "):
            groups += [ g.strip() for g in r[6:].split(",")]

    if groups:
        try:
            res = proxy.check_membership(req.user, groups)
        except Exception, e:
            req.log_error('PYNTLM: Unexpected error when checking membership of %s in groups %s for URI %s: %s' % (req.user,str(groups),req.unparsed_uri,str(e)))
        if res:
            #req.log_error('PYNTLM: Groups before %s' % str(cacheGroups._cache))
            cacheGroups.add(rules, req.user)
            #req.log_error('PYNTLM: Groups after %s' % str(cacheGroups._cache))
            req.log_error('PYNTLM: Membership check succeeded for %s in groups %s for URI %s.' %
                (req.user,str(groups),req.unparsed_uri), apache.APLOG_INFO)
            return True
        req.log_error('PYNTLM: Membership check failed of %s in groups %s for URI %s.' %
            (req.user,str(groups),req.unparsed_uri))
    else:
        req.log_error('PYNTLM: Authorization failed for %s and URI %s.' %
            (req.user,req.unparsed_uri))
    return False

def handle_type3(req, ntlm_message):
    '''Handle a Type3 NTLM message. Send it to the Domain Controller
    and get back the final authentication outcome.

    @req            The request that carried the message
    @ntlm_message   The actual Type3 message, in binary format
    '''
    
    proxy = cache.get_proxy(req.connection.id)
    try:
        user, domain = parse_ntlm_authenticate(ntlm_message)
        result = proxy.authenticate(ntlm_message)
    except Exception, e:
        req.log_error('PYNTLM: Error when retrieving Type 3 message from server = %s' % str(e), apache.APLOG_CRIT)
        user, domain = 'invalid', 'invalid'
        result = False
    if not result:
        cache.remove(req.connection.id)
        req.log_error('PYNTLM: User %s/%s at %s failed authentication for URI %s' % (
            domain,user,req.connection.remote_ip,req.unparsed_uri))
        return handle_unauthorized(req)

    req.log_error('PYNTLM: User %s/%s has been authenticated to access URI %s' % (user,domain,req.unparsed_uri), apache.APLOG_NOTICE)
    req.user = user
    result = check_authorization(req, proxy)
    cache.remove(req.connection.id)

    if not result:
        return apache.HTTP_FORBIDDEN

    req.connection.notes.add('NTLM_AUTHORIZED',user)
    return apache.OK
    
def handle_basic(req, user, password):
    '''Handle a request authenticated using the Basic Access Authentication
    mechanism (RFC2617).
    '''
    req.log_error('Handling Basic Access Authentication for URI %s' % (req.unparsed_uri))

    domain = req.get_options().get('Domain', req.auth_name())
    client = NTLM_Client(user, domain, password)
    type1 = client.make_ntlm_negotiate()

    try:
        (proxy, type2) = connect_to_proxy(req, type1)
    except Exception, e:
        return apache.HTTP_INTERNAL_SERVER_ERROR
    
    client.parse_ntlm_challenge(type2)
    type3 = client.make_ntlm_authenticate()
    if not proxy.authenticate(type3):
        proxy.close()
        req.log_error('PYNTLM: User %s/%s at %s failed Basic authentication for URI %s' % (
            domain,user,req.connection.remote_ip,req.unparsed_uri))
        return handle_unauthorized(req)
    
    req.log_error('PYNTLM: User %s/%s has been authenticated (Basic) to access URI %s' % (user,domain,req.unparsed_uri), apache.APLOG_NOTICE)
    req.user = user
    result = check_authorization(req, proxy)
    proxy.close()

    if not result:
        return apache.HTTP_FORBIDDEN

    req.connection.notes.add('BASIC_AUTHORIZED', user+password)
    return apache.OK
    
def authenhandler(req):
    '''The request handler called by mod_python in the authentication phase.'''
    req.log_error("PYNTLM: Handling connection 0x%X from address %s for %s URI %s. %d entries in connection cache." % (
        req.connection.id, req.connection.remote_ip,req.method,req.unparsed_uri,len(cache)), apache.APLOG_INFO)

    # Extract Authorization header, as a list (if present)
    auth_headers = req.headers_in.get('Authorization', [])
    if not isinstance(auth_headers, list):
        auth_headers = [ auth_headers ]

    # If this connection was authenticated with NTLM, quit immediately with an OK
    # (unless it comes from IE).
    user = req.connection.notes.get('NTLM_AUTHORIZED', None)
    if user:
        req.user = user
        # Internet Explorer sends a Type 1 authorization request with an empty
        # POST, even if the connection is already authenticated.
        # We don't de-authenticate the user (meaning that we don't remove the
        # NTLM_AUTHORIZED key from connection.notes), but we still let a new
        # challenge-response exchange take place.
        # For other methods, it is acceptable to return OK immediately.
        if  auth_headers:
            req.log_error('PYTNLM: Spurious authentication request on connection 0x%X. Method = %s. Content-Length = %d. Headers = %s' % (
            req.connection.id, req.method, req.clength, auth_headers), apache.APLOG_INFO)
            if req.method!='POST' or req.clength>0:
                return apache.OK
        else:
            return apache.OK
    
    # If there is no Authorization header it means it is the first request.
    # We reject it with a 401, indicating which authentication protocol we understand.
    if not auth_headers:
        return handle_unauthorized(req)

    # Extract authentication data from any of the Authorization headers
    try:
        for ah in auth_headers:
            ah_data = decode_http_authorization_header(ah)
            if ah_data:
                break
    except:
        ah_data = False
    
    if not ah_data:
        req.log_error('Error when parsing Authorization header from address %s and URI %s' % (
        req.connection.remote_ip,req.unparsed_uri), apache.APLOG_ERR)
        return apache.HTTP_BAD_REQUEST

    if ah_data[0]=='Basic':
        # If this connection was authenticated with Basic, verify that the
        # credentials match and return 200 (if they do) or 401 (if they
        # don't). We don't need to actually query the DC for that.
        userpwd = req.connection.notes.get('BASIC_AUTHORIZED', None)
        if userpwd:
            if userpwd != ah_data[1]+ah_data[2]:
                return handle_unauthorized(req)
            req.user = ah_data[1]
            return apache.OK
        # Connection was not authenticated before
        return handle_basic(req, ah_data[1], ah_data[2])

    # If we get here it means that there is an Authorization header, with an
    # NTLM message in it. Moreover, the connection needs to be (re)authenticated.
    # Most likely, the NTLM message is of:
    # - Type 1 (and there is nothing in the cache): the client wants to
    #   authenticate for the first time,
    # - Type 3 (and there is something in the cache): the client wants to finalize
    #   a pending authentication request.
    #
    # However, it could still be that there is a Type 3 and nothing in the
    # cache (the final client message was erroneously routed to a new connection),
    # or that there is a Type 1 with something in the cache (the client wants to
    # initiate an cancel a pending authentication).

    try:
        ntlm_version = ntlm_message_type(ah_data[1])
        if ntlm_version==1:
            return handle_type1(req, ah_data[1]) 
        if ntlm_version==3:
            if cache.has_key(req.connection.id):
                return handle_type3(req, ah_data[1])
            req.log_error('Unexpected NTLM message Type 3 in new connection from address %s and URI %s' %
                (req.connection.remote_ip,req.unparsed_uri), apache.APLOG_INFO)
            return handle_unauthorized(req)
        error = 'Type 2 message in client request'
    except Exception, e:
        error = str(e)
    req.log_error('Incorrect NTLM message in Authorization header from address %s and URI %s: %s' %
            (req.connection.remote_ip,req.unparsed_uri,error), apache.APLOG_ERR)
    return apache.HTTP_BAD_REQUEST

